clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -737,14 +737,25 @@ public synchronized void maybeTransitionToErrorState(RuntimeException exception)
     }
 
     synchronized void handleFailedBatch(ProducerBatch batch, RuntimeException exception, boolean adjustSequenceNumbers) {
-        maybeTransitionToErrorState(exception);
+        // Compare the batch with the current ProducerIdAndEpoch. If the producer IDs are the *same* but the epochs
+        // are *different*, consider the batch as stale.
+        boolean isStaleBatch = batch.producerId() == producerIdAndEpoch.producerId && batch.producerEpoch() != producerIdAndEpoch.epoch;
+
+        if (!isStaleBatch && !hasFatalError())
+            maybeTransitionToErrorState(exception);
```

Comment:
Is this comment helpful or distracting?

```suggestion
        // It's possible the transaction manager is already in the FATAL_ERROR
        // state at this point. Depending on the incoming exception type,
        // maybeTransitionToErrorState() could attempt to set the state to
        // ABORTABLE_ERROR. For example, suppose a fatal error occurred during a
        // transaction, and then moments later one of the batches in that
        // transaction failed with a TimeoutException.
        // maybeTransitionToErrorState() would then (blindly) attempt to
        // transition to ABORTABLE_ERROR, which is invalid and would result
        // in an IllegalStateException.
        //
        // Therefore, only attempt to transition to the FATAL_ERROR state if
        // the batch is "fresh" *and* the transaction manager is not already
        // in the FATAL_ERROR state.
        if (!isStaleBatch && !hasFatalError())
            maybeTransitionToErrorState(exception);
```

Reply:
```
        // Therefore, only attempt to transition to the FATAL_ERROR state if
        // the batch is "fresh" *and* the transaction manager is not already
        // in the FATAL_ERROR state.
```

FATAL_ERROR to FATAL_ERROR[ is not an invalid transition](https://github.com/k-raina/kafka/blob/ab2cae14b93c13e21eda1d142d582bb7f42d7aec/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L211-L233). Do you mean  `only attempt transition to the ABORTABLE_ERROR state`

Reply:
@k-raina Did you have a suggestion for better wording? Happy to hear it 😄

Reply:
Comment seems fine to me.